login-utils/nologin.8
.TH NOLOGIN 8 "November 2019" "util-linux" "System Administration"
.SH NAME
nologin \- politely refuse a login
.SH SYNOPSIS
.B nologin
.RB [ \-V ]
.RB [ \-h ]
.SH DESCRIPTION
.B nologin
displays a message that an account is not available and exits non-zero.  It is
intended as a replacement shell field to deny login access to an account.
.PP
If the file
.IR /etc/nologin.txt exists ,
.B nologin
displays its contents to the
user instead of the default message.
.PP
The exit status returned by
.B nologin
is always 1.
.SH OPTIONS
\fB\-c\fR, \fB\-\-command\fR \fIcommand\fR
.br
\fB\-\-init-file\fR
.br
\fB\-i\fR \fB\-\-interactive\fR
.br
\fB\-\-init-file\fR \fIfile\fR
.br
\fB\-i\fR, \fB\-\-interactive\fR
.br
\fB\-l\fR, \fB\-\-login\fR
.br
\fB\-\-noprofile\fR
.br
\fB\-\-norc\fR
.br
\fB\-\-posix\fR
.br
\fB\-\-rcfile\fR \fIfile\fR
.br
\fB\-r\fR, \fB\-\-restricted\fR
.IP
These shell command-line options are ignored to avoid nologin error.
.IP "\fB\-h\fR, \fB\-\-help\fR"
Display help text and exit.
.IP "\fB\-V\fR, \fB\-\-version\fR"
Display version information and exit.
.SH NOTES
.B nologin
is a per-account way to disable login (usually used for system accounts like http or ftp).
.BR nologin (8)
uses
.I /etc/nologin.txt
as an optional source for a non-default message, the login
access is always refused independently of the file.
.PP
.BR pam_nologin (8)
PAM module usually prevents all non-root users from logging into the system.
.BR pam_nologin (8)
functionality is controlled by
.I /var/run/nologin
or the
.I /etc/nologin
file.
.SH HISTORY
The
.B nologin
command appeared in 4.4BSD.
.SH AUTHORS
.UR kzak@redhat.com
Karel Zak
.UE
.SH SEE ALSO
.BR login (1),
.BR passwd (5),
.BR pam_nologin (8)
.SH AVAILABILITY
The nologin command is part of the util-linux package and is available from
.UR https://\:www.kernel.org\:/pub\:/linux\:/utils\:/util-linux/
Linux Kernel Archive
.UE .


NEWS
util-linux 2.36: Jul 23 2020
* see Documentation/releases/v2.36-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36-ChangeLog

util-linux 2.36-rc2: Jul 09 2020
* see Documentation/releases/v2.36-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36-rc2-ChangeLog

util-linux 2.36-rc1: Jun 09 2020
* see Documentation/releases/v2.36-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36-rc1-ChangeLog

util-linux 2.35: Jan 21 2020
* see Documentation/releases/v2.35-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.35/v2.35-ChangeLog

util-linux 2.35-rc2: Jan 8 2020
* see Documentation/releases/v2.35-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.35/v2.35-rc2-ChangeLog

util-linux 2.35-rc1: Dec 11 2019
* see Documentation/releases/v2.35-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.35/v2.35-rc1-ChangeLog

util-linux 2.34: Jun 14 2019
* see Documentation/releases/v2.34-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.34/v2.34-ChangeLog

util-linux 2.34-rc2: May 30 2019
* see Documentation/releases/v2.34-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.34/v2.34-rc2-ChangeLog

util-linux 2.34-rc1: Apr 30 2019
* see Documentation/releases/v2.34-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.34/v2.34-rc1-ChangeLog

util-linux 2.33: Nov 6 2018
* see Documentation/releases/v2.33-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-ChangeLog

util-linux 2.33-rc2: Oct 19 2018
* see Documentation/releases/v2.33-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-rc2-ChangeLog

util-linux 2.33-rc1: Sep 25 2018
* see Documentation/releases/v2.33-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-rc1-ChangeLog

util-linux 2.32: Mar 21 2018
* see Documentation/releases/v2.32-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.32/v2.32-ChangeLog

util-linux 2.32-rc2: Mar 01 2018
* see Documentation/releases/v2.32-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.32/v2.32-rc2-ChangeLog

util-linux 2.32-rc1: Feb 13 2018
* see Documentation/releases/v2.32-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.32/v2.32-rc1-ChangeLog

util-linux 2.31: Oct 19 2017
* see Documentation/releases/v2.31-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.31/v2.31-ChangeLog

util-linux 2.31-rc2: Oct 03 2017
* see Documentation/releases/v2.31-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.31/v2.31-rc2-ChangeLog

util-linux 2.31-rc1: Sep 22 2017
* see Documentation/releases/v2.31-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.31/v2.31-rc1-ChangeLog

util-linux 2.30: Jun 02 2017
* see Documentation/releases/v2.30-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.30/v2.30-ChangeLog

util-linux 2.30-rc2: May 12 2017
* see Documentation/releases/v2.30-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.30/v2.30-rc2-ChangeLog

util-linux 2.30-rc1: May 12 2017
* see Documentation/releases/v2.30-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.30/v2.30-rc1-ChangeLog

util-linux 2.29: Nov 8 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-ChangeLog

util-linux 2.29-rc2: Oct 19 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-rc2-ChangeLog

util-linux 2.29-rc1: Sep 30 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-rc1-ChangeLog

util-linux 2.28: Apr 12 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-ChangeLog

util-linux 2.28-rc2: Mar 29 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-rc2-ChangeLog

util-linux 2.28-rc1: Mar 11 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-rc1-ChangeLog

util-linux 2.27: Sep 07 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-ChangeLog

util-linux 2.27-rc2: Aug 24 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-rc2-ChangeLog

util-linux 2.27-rc1: Jul 31 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-rc1-ChangeLog

util-linux 2.26: Feb 19 2015
* see Documentation/releases/v2.26-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.26/v2.26-ChangeLog

util-linux 2.26-rc2: Feb 4 2015
* see Documentation/releases/v2.26-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.26/v2.26-rc2-ChangeLog

util-linux 2.26-rc1: Jan 14 2015

login-utils/nologin.c
/*
 * Copyright (C) 2013 Karel Zak <kzak@redhat.com>
 */

#include <stdio.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <getopt.h>

#include "c.h"
#include "nls.h"
#include "pathnames.h"

/*
 * Always return EXIT_FAILURE (1), don't try to be smart!
 */

static void __attribute__((__noreturn__)) usage(void)
{
	FILE *out = stdout;
	fputs(USAGE_HEADER, out);
	fprintf(out,
		_(" %s [options]\n"), program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Politely refuse a login.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -c, --command <command>  does nothing (for compatibility with su -c)\n"), out);
	printf(USAGE_HELP_OPTIONS(26));

	printf(USAGE_MAN_TAIL("nologin(8)"));
	exit(EXIT_FAILURE);
}

int main(int argc, char *argv[])
{
	int c, fd = -1;
	struct stat st;
	enum {
		OPT_INIT_FILE = CHAR_MAX + 1,
		OPT_NOPROFILE,
		OPT_NORC,
		OPT_POSIX,
		OPT_RCFILE
	};
	static const struct option longopts[] = {
		{ "command",     required_argument, NULL, 'c'           },
		{ "init-file",   required_argument, NULL, OPT_INIT_FILE },
		{ "interactive", no_argument,       NULL, 'i'           },
		{ "login",       no_argument,       NULL, 'l'           },
		{ "noprofile",   no_argument,       NULL, OPT_NOPROFILE },
		{ "norc",        no_argument,       NULL, OPT_NORC      },
		{ "posix",       no_argument,       NULL, OPT_POSIX     },
		{ "rcfile",      required_argument, NULL, OPT_RCFILE    },
		{ "restricted",  no_argument,       NULL, 'r'           },
		{ "help",        no_argument,       NULL, 'h'           },
		{ "version",     no_argument,       NULL, 'V'           },
		{ NULL, 0, NULL, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	while ((c = getopt_long(argc, argv, "c:ilrhV", longopts, NULL)) != -1) {
		switch (c) {
		case 'c':
		case OPT_INIT_FILE:
		case 'i':
		case 'l':
		case OPT_NOPROFILE:
		case OPT_NORC:
		case OPT_POSIX:
		case OPT_RCFILE:
		case 'r':
			/* Ignore well known shell command-line options */
			break;
		case 'h':
			usage();
		case 'V':
			print_version(EXIT_FAILURE);	/* yes FAILURE! */
		default:
			errtryhelp(EXIT_FAILURE);
		}
	}

	fd = open(_PATH_NOLOGIN_TXT, O_RDONLY);
	if (fd < 0)
		goto dflt;

	c = fstat(fd, &st);
	if (c < 0 || !S_ISREG(st.st_mode))
		goto dflt;
	else {
		char buf[BUFSIZ];
		ssize_t rd;

		while ((rd = read(fd, buf, sizeof(buf))) > 0)
			ignore_result( write(STDOUT_FILENO, buf, rd) );

		close(fd);
		return EXIT_FAILURE;
	}

dflt:
	if (fd >= 0)
		close(fd);
	fprintf(stdout, _("This account is currently not available.\n"));
	return EXIT_FAILURE;
}
